app/models/reservation.rb

```diff
@@ -189,13 +189,32 @@ def can_cancel?
   end
 
   def can_customer_edit?
-    !canceled? && !complete? && before_lock_window?
+    !canceled? && !complete? && (reserve_start_at_editable? || reserve_end_at_editable?)
+  end
+
+  def reserve_start_at_editable?
+    before_lock_window? && actual_start_at.blank?
+  end
+
+  def reserve_end_at_editable?
+    outside_lock_window? && Time.zone.now <= reserve_end_at && next_duration_available? && actual_end_at.blank?
+  end
+
+  def next_duration_available?
```

Comment:
Is this at all related to `ReservationFinder#next_reservation` in Product::SchedulingSupport? It's a really confusing method, but it might do something similar to what you're looking for.

We should make sure that you also can't extend your reservation outside of the schedule rules. E.g. There's a schedule rule for 9-5, and you have a reservation from 4-5. You shouldn't be able to extend beyond that.